core/capabilities/vault/capability.go

```diff
@@ -19,25 +19,57 @@ import (
 	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/requests"
 	jsonrpc "github.com/smartcontractkit/chainlink-common/pkg/jsonrpc2"
 	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/core"
 	"github.com/smartcontractkit/chainlink/v2/core/capabilities/vault/vaulttypes"
 )
 
 var _ capabilities.ExecutableCapability = (*Capability)(nil)
 
 type Capability struct {
-	lggr              logger.Logger
-	clock             clockwork.Clock
-	expiresAfter      time.Duration
-	handler           *requests.Handler[*vaulttypes.Request, *vaulttypes.Response]
-	requestAuthorizer RequestAuthorizer
-	publicKey         *LazyPublicKey
+	lggr                 logger.Logger
+	clock                clockwork.Clock
+	expiresAfter         time.Duration
+	handler              *requests.Handler[*vaulttypes.Request, *vaulttypes.Response]
+	requestAuthorizer    RequestAuthorizer
+	capabilitiesRegistry core.CapabilitiesRegistry
+	publicKey            *LazyPublicKey
 }
 
 func (s *Capability) Start(ctx context.Context) error {
-	return s.handler.Start(ctx)
+	if err := s.handler.Start(ctx); err != nil {
+		return fmt.Errorf("error starting vault DON request handler: %w", err)
+	}
+
+	closeHandler := func() {
+		ierr := s.handler.Close()
+		if ierr != nil {
+			s.lggr.Errorf("error closing vault DON request handler after failed registration: %v", ierr)
+		}
+	}
+
+	err := s.capabilitiesRegistry.Add(ctx, s)
+	if err != nil {
+		closeHandler()
+		return fmt.Errorf("error registering vault capability: %w", err)
+	}
+
+	return nil
 }
 
 func (s *Capability) Close() error {
+	closeHandler := func() {
+		ierr := s.handler.Close()
+		if ierr != nil {
+			s.lggr.Errorf("error closing vault DON request handler after failed unregistration: %v", ierr)
+		}
+	}
+
+	err := s.capabilitiesRegistry.Remove(context.Background(), vaultcommon.CapabilityID)
+	if err != nil {
+		closeHandler()
+		return fmt.Errorf("error unregistering vault capability: %w", err)
+	}
+
 	return s.handler.Close()
```

Comment:
How about joining the errors?
```suggestion
func (s *Capability) Close() (errs error) {
	err := s.capabilitiesRegistry.Remove(context.Background(), vaultcommon.CapabilityID)
	if err != nil {
		errs = errors.Join(errs, fmt.Errorf("error unregistering vault capability: %w", err))
	}
	err = s.handler.Close()
	if err != nil {
		errs = errors.Join(errs, fmt.Errorf("error closing vault DON request handler after failed unregistration: %v", err))
	}

	return
```

Reply:
Yeah this is a better suggestion! Will change it to this